Design/histogram_type.md

```diff
@@ -0,0 +1,108 @@
+# Histogram Data Type
+
+**Note:** This document is mainly concerned with a data type for a *generic* histogram.
+In the context considered here, a histogram is *not* specific to neutron science and probably not even specific to physics.
+In particular it should not be confused with the current data types in Mantid that are joint containers for histogram data and spectrum/detector information, such as `ISpectrum` and `Histogram1D`.
+
+
+## Motivation
+
+One of the most fundamental concepts of ubiquitous use in Mantid is that of histograms of (event) data.
+Despite this, there is currently no data type in Mantid that deals with histogram data.
+The purpose of this design proposal is to fix this.
+
+There are various motivations for this:
+
+- Reduce code duplication by providing arithmetic operations such as addition of histograms.
+- Hide handling of errors (uncertainties).
+- Hide discrimination between histogram and distribution data.
+- Decrease risk for certain types of bugs by using the C++ type system.
+
+All this would remove duplicated code from the client and put it into a central place where it can be unit-tested.
+
+
+## Current situation
+
+The current way of dealing with histograms has various shortcomings.
+Basically we have a set of three `std::vector<double>`, for `X`, `Y`, and `E`.
+There is a single `typedef` `MantidVec` used for all three of them (and other data), and this is probably adding only more confusion.
+
+The problems include:
+
+- No type safety:
+  - The bug in `ConvertUnits` creped in because we are currently using the same instance of a single type for both histogram data and distribution data. The actual meaning of a vector of `Y` values is basically kept as a mental flag in the brain of a Mantid developer.
+  - We are using a single type for `X`, `Y`, and `E`. It is thus easy to accidentally mix them up.
+  - The array `E` of error values is used for both the variance and the standard deviation. The state is again carried as a flag in the brain of a developer. We even have functions such as `rebin` that return an array of `E` values that may mean either variance or standard deviation depending on a flag to the function!
+- Code duplication: Re-implementing the same thing 10 times is wasteful and error prone.
+  - Operations like adding histograms are frequent.
```

Comment:
According to my list, I so far found the following:

- assignment
- set to zero
- arithmetics with other vectors, including `+`, `+=`, etc.
- arithmetics with a scalar, e.g., adding a constant to the bin-edge vector or the bin values
- comparison (`==`)
- integration
- rebin
- reverse order
- apply function to values (similar to `std::transform`)
- truncate head or tail
- get lambda (i.e., compute bin centers)